configuration-versioning/README.md

```diff
@@ -0,0 +1,141 @@
+      owner: packit
+      project: packit-dev
+```
+
+and schemas:
+
+```python
+class ImageBuilderMetadataV1(ImageBuilderMetadataSchema):
+    image_distribution = fields.String(missing=None)
+    image_request = fields.Dict(missing=None)
+    image_customizations = fields.Dict(missing=None)
+```
+
+```python
+class ImageBuilderMetadataV2(ImageBuilderMetadataSchema):
+    image_distribution = fields.String(missing=None)
+    image_type: fields.String(missing=None)
+    image_architecture = fields.String(missing=None)
+    image_account_id = fields.String(missing=None)
+    packages_to_install = fields.String(missing=None)
+    owner = fields.String(missing=None)
+    project = fields.String(missing=None)
+```
+
+and the config object would have all the fields and the code could handle both
+
+- from the coding perspective this would not be a problem - the version would be detected either from whole config
+  or directly from the `image_builder_metadata`
+- but this may be confusing to users - versioning some configuration field apart from the whole config,
+  on the other hand it would be independent on the rest of the package config
+
+## Implementation details
+
+- all versions of the configuration need to be migrated when loading the configuration to the latest state
+  of the configuration object
+
+### Versioning
+
+- for each backwards incompatible change, we would bump the version
+- version would default to 1 (if not present)
+- this would be properly documented, we can get inspired by Docker,
+  [see](https://docs.docker.com/compose/compose-file/compose-versioning/#versioning), the documentation
+  would also include Packit version that starts supporting the particular configuration schema
+
+### Having more schemas vs having one schema
+
+#### Having more schemas
+
+- maintaining more versions of schema for whole package config should not be needed
+  - the schemas would mostly have the same content, only with differences in a few fields
+- this approach could be used for smaller parts of the configuration, e.g.
+  creating nested schema (as described for Example 1)
+- the coding around would be pretty straightforward, detect the version
+  and use the correct schema for loading/dumping accordingly
+
+#### Having one schema
+
+- in case of Example 2, the schema itself doesn't change
+- we could use the `pre_load` and `pre_dump` (we already use these) to migrate the data:
+
+```python
+    @staticmethod
+    def propagate_options_to_jobs(data, **_):
+        """
+        add all the fields (except for jobs) to every job so we can process only jobconfig in p-s
+        """
+        if not data:  # data is None when .packit.yaml is empty
+            return data
+        for job in data.get("jobs", []):
+            for k, v in data.items():
+                if k == "jobs":
+                    # overriding jobs doesn't make any sense
+                    continue
+                if self.version == 1:
```

Comment:
Doing this in the `pre_load` step seems late to me, and will make code more and more complicated over time.

I would rather do something like: check the version in `PackageConfig.get_from_dict()` (default to some pre-defined version if the version is not specified – the last one could be a good candidate) and then import the schema classes matching that version:

```python
if version == "1":
    from packit.schema.v1 import PackageConfigSchema
elif version == "2":
    from packit.schema.v2 import PackageConfigSchema
```

This would allow partially redefining schema classes or reexporting existing schema classes. For example in `packit.schema.v2` one could have:

```python
from packit.schema.v1 import PackageConfigSchema as PackageConfigSchemaV1
# reexport as v2, without a change
from packit.schema.v1 import JobConfig

class PackageConfigSchema(PackageConfigSchemaV1):
    # override some fields
```

Would something like this make sense?

Reply:
yes, this makes sense, I will incorporate it into the document